### ISSUE TITLE
TypeError: state is not a function

### ISSUE THREAD

**Issue by Victorystick** [Rich-Harris/tippex#1]
I'm using `tippex` in https://github.com/rollup/rollup-plugin-typescript, and encountered this error while attempting to bundle Angular2's TypeScript sources. I know it wasn't intended to be used on TypeScript, so the difference in grammars might explain the crash.

```
state is not a function
TypeError: state is not a function
    at find (/Users/oskar/github/rollup-plugin-typescript/node_modules/tippex/www/tippex/src/index.js:134:11)
    at Object.erase (/Users/oskar/github/rollup-plugin-typescript/node_modules/tippex/www/tippex/src/index.js:156:16)
    at fix (/Users/oskar/github/rollup-plugin-typescript/dist/rollup-plugin-typescript.cjs.js:47:28)
    at transform (/Users/oskar/github/rollup-plugin-typescript/dist/rollup-plugin-typescript.cjs.js:122:58)
    at /Users/oskar/github/rollup/Users/oskar/github/rollup/src/utils/transform.js:18:28
    at tryCatch (/Users/oskar/github/rollup/Users/oskar/github/rollup/node_modules/es6-promise/lib/es6-promise/-internal.js:182:12)
    at invokeCallback (/Users/oskar/github/rollup/Users/oskar/github/rollup/node_modules/es6-promise/lib/es6-promise/-internal.js:194:13)
    at /Users/oskar/github/rollup/Users/oskar/github/rollup/node_modules/es6-promise/lib/es6-promise/promise.js:376:9
    at flush (/Users/oskar/github/rollup/Users/oskar/github/rollup/node_modules/es6-promise/lib/es6-promise/asap.js:88:5)
    at doNTCallback0 (node.js:428:9)
```

**Comment by pkozlowski-opensource** [comment]
Just hitting it myself while trying to bundle Angular2 from TypeScript sources. Will investigate.

**Comment by Rich-Harris** [comment]
This may have been related to #2 – I've just released 2.0.0 which might fix the issue. Can you give it a go please? (Sorry for the long radio silence)

**Comment by RReverser** [comment]
I found a repro case that leads to this error, too (not sure if the same):

``` javascript
const stuff = `$`;
```

Workaround is escaping the `$` sign, then `tippex` works fine:

``` javascript
const stuff = `\$`;
```

Given this, I suspect that tippex expects `$` to be always part of `${...}` substitution unless escaped, which is not necessarily true.

**Comment by Victorystick** [comment]
My original issue was resolved by #2 and #7, but I encountered another error with the same symptoms; i.e. `state is not a function`.

Tippex doesn't handle pairs of trailing asterisks in block comments. The logical error in the implementation was the same as in #7. Patch is available as of #10.

**Comment by RReverser** [comment]
Why isn't `tippex` just using a proper JS tokenizer anyway? (like one from Acorn)
Should be much easier to support all the syntax.

**Comment by Victorystick** [comment]
@RReverser I assume it's to reduce complexity and avoid dependencies; although I understand what you mean. It's battle tested for sure.

It suits me well since I'm using it on TypeScript code which Acorn's tokenizer would reject. Or perhaps it would just tokenize all those TypeScript keywords as identifiers. 💭

**Comment by RReverser** [comment]
@Victorystick I can't think of any TS code which would contain invalid JS tokens off the head. Can you please provide any example?

**Comment by Victorystick** [comment]
@RReverser No, you're probably right. 😃

**Comment by Rich-Harris** [comment]
> Why isn't `tippex` just using a proper JS tokenizer anyway? (like one from Acorn)

It's designed to be a very fast, very lightweight tool for stripping comments and strings so that it's easy to find certain patterns (e.g. `import` or `require` statements) – original use case was for parsing [Ractive](http://www.ractivejs.org/) component definitions in the browser, where shipping something like Acorn isn't really appropriate.

https://github.com/ternjs/acorn/blob/master/src/tokenize.js alone is more than twice the size of the whole of tippex. So I'd prefer to keep pushing forward with fixing bugs in tippex rather than switching, even if I am a big Acorn fan as you know 😀

**Comment by RReverser** [comment]
@Rich-Harris Unless after fixing all bugs you will end up with the same size :D

**Comment by Rich-Harris** [comment]
Released 2.1.2 with #10 – will be optimistic and close this!

**Comment by bcherny** [comment]
Still happening for me with 2.1.2. Repro code:

``` sh
echo "export default /^[-a-z0-9~\!\$%^&*_=+}{\'?]+(\.[-a-z0-9~!$%^&*_=+}{\'?]+)*@([a-z0-9_][-a-z0-9_]*(\.[-a-z0-9_]+)*\.([a-z]{2,10})|([0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}))(:[0-9]{1,5})?$/i" | rollup -c
```

**Comment by Rich-Harris** [comment]
Thanks @bcherny – have released 2.2.0 which fixes this (it wasn't expecting a regex literal because it didn't know how to handle `export default`, and so it fell apart at the first `}` character, thinking it closed a block that wasn't open).

I've also made the error message rather more informative than 'state is not a function', so that any similar bugs in future are easier to pin down.
